### ISSUE TITLE
Support Cosign keyless signing & verification

### ISSUE THREAD

**Issue by developer-guy** [goreleaser/goreleaser#2659]
**Is your feature request related to a problem? Please describe.**
It is not related to a problem. It's a feature request for signing and verifying with Keyless Mode.[^1]

**Describe the solution you'd like**

GoReleaser can use the `cosign` tool as a command in the `sigs` section.  In order to enable Keyless mode in `cosign`, we need to pass an environment variable called `COSIGN_EXPERIMENTAL=1`. So, we've opened another issue to enable passing environment variables to the `sigs` section. Also, recently, the Kyverno project did the same thing. [^2]

[ ] - https://github.com/goreleaser/goreleaser/issues/2657 - @Dentrax 

**Describe alternatives you've considered**

**Additional context**
[^1]: https://github.com/sigstore/cosign/blob/main/KEYLESS.md
[^2]: github.com/kyverno/kyverno/pull/2677

cc: @dentrax @erkanzileli

**Comment by developer-guy** [comment]
here is an example for it, seems it's using the keyless mode:

https://github.com/in-toto/in-toto-golang/pull/128/files

**Comment by shibumi** [comment]
@developer-guy yes, this is correct. I specify the `COSIGN_EXPERIMENTAL=1` variable in my Github Actions Workflow.
Also, make sure to give the github actions workflow enough permissions. You will need: `id-token: write` and `contents: write`.

I hope the cosign_experimental var is not necessary anymore in the future... the sigstore team is working towards stable features right now.

**Comment by lukehinds** [comment]
> I hope the cosign_experimental var is not necessary anymore in the future... the sigstore team is working towards stable features right now.

As soon as we hit 1.0 in rekor, it's out of experimental. The main aspect for 1.0 is an API change and an operational procedure to shard the tlog.

**Comment by developer-guy** [comment]
@shibumi can you explain a bit about how you can verify-blob once you sign the blob? On locally, we got the certificate from the output of the sign-blob command, then used it with the `--cert` flag of the verify-blob command while verifying the signature. In GoReleaser flow, how can we get the certificate for later use of verifying signature?

cc: @dentrax @dlorenc @imjasonh @hectorj2f

**Comment by developer-guy** [comment]
> @shibumi can you explain a bit about how you can verify-blob once you sign the blob? On locally, we got the certificate from the output of the sign-blob command, then used it with the `--cert` flag of the verify-blob command while verifying the signature. In GoReleaser flow, how can we get the certificate for later use of verifying signature?
> 
> cc: @Dentrax @dlorenc @imjasonh @hectorj2f

According to this, we've opened a PR to output the certificate, which we can use while verifying the signature; please [see](github.com/sigstore/cosign/pull/1021). Once it is merged, eventually, we can add this short-lived certificate to the releases. But we are not sure about it because once this certificate gets expired it might be a problem for the end-users. So, this is another topic that we should consider.  Maybe @lukehinds @dlorenc can enlighten us about what we should do.

**Comment by Dentrax** [comment]
> According to this, we've opened a PR to output the certificate, which we can use while verifying the signature; please [see](github.com/sigstore/cosign/pull/1021).

Eventually, we will be able to pass this _certificate_ file to goreleaser _artifacts_, and expect 3 files after each release to use later using cosign to verify: `checksum.txt`, `checksum.txt.sig`, `cosign.cert`.

**Comment by shibumi** [comment]
> @shibumi can you explain a bit about how you can verify-blob once you sign the blob? On locally, we got the certificate from the output of the sign-blob command, then used it with the --cert flag of the verify-blob command while verifying the signature. In GoReleaser flow, how can we get the certificate for later use of verifying signature?

Hi @developer-guy, with GoReleaser it's a little bit difficult. I have created a PR for cosign that will allow storing the generated public certificate on disk for OIDC-issuers: https://github.com/sigstore/cosign/pull/1016

I just realized that you have created a PR for this, too :D Two people with thesame idea.

The next cosign release will automatically retrieve the certificate via searching through rekor with the signature (I think).
But Dan Lorenc told me, having the certificate on disk is a nice addition.

**Comment by shibumi** [comment]
> Eventually, we will be able to pass this certificate file to goreleaser artifacts, and expect 3 files after each release to use later using cosign to verify: checksum.txt, checksum.txt.sig, cosign.cert.

I would suggest something like this:
```yaml
signs:
  - cmd: cosign
    signature: "${artifact}.sig"
    certificate: "${artifact.pem"
    args: ["sign-blob", "-oidc-issuer=https://token.actions.githubusercontent.com", "-output=${signature}", "-output-pem": "${certificate}, "${artifact}"]
```

**Comment by developer-guy** [comment]
>  certificate: "${artifact}.pem"

GoReleaser does not support it yet. Should we open a PR to enable this @shibumi, or did you already work on it 😜

**Comment by shibumi** [comment]
> GoReleaser does not support it yet. Should we open a PR to enable this @shibumi, or did you already work on it @shibumi stuck_out_tongue_winking_eye

For GoReleaser I do not have a PR. Feel free to open one :)

**Comment by caarlos0** [comment]
so, to summarize the idea:

- we add a "certificate" env/var to signs
  - possibly also allow the user to name this file
- cosign writes the key to that file
- goreleaser should add this file to its artifact list and publish it on the release 

if that's it, I have an idea on how to impl it, can do it later today

**Comment by shibumi** [comment]
@caarlos0 Yep, this is correct.

**Comment by developer-guy** [comment]
> so, to summarize the idea:
> 
> * we add a "certificate" env/var to signs
>   
>   * possibly also allow the user to name this file
> * cosign writes the key to that file
> * goreleaser should add this file to its artifact list and publish it on the release
> 
> if that's it, I have an idea on how to impl it, can do it later today

Exactly! But let's wait for the maintainer's thoughts first; if they agree on this idea, we can start to work on this; we are also really excited about the developments in this area; we also want to develop this feature if you let us do that.

**Comment by dirien** [comment]
Totaly love this feature! Going to be awsome! thx @developer-guy, @Dentrax and @shibumi! 👍🏻👍🏻👍🏻

**Comment by lukehinds** [comment]
Here is the deal with keyless and certificate expiry, fulcio certs are designed to be very short lived by design (I think its currently 30 seconds).

We go through the following flow:

user generates key pair -> contacts fulcio  -> fulcio provides them with an OpenID session (login with google, github etc) -> user also sends a challenge (there email) encrypted with their key pair. 

Fulcio then requests the email address of the user from the identity provider (this is granted via the OpenID connect flow). It then places email into the x509 certificate (Subject Alternative Name) along with the public key from the key pair used to send the challenge (proving the user has the key pair in their possession. This certificate is also stored into the transparency log.

The user then signs their blob using signature and stores in rekor the digest / signature and the x509 signing certificate. 

After reaching this stage, they can discard the private key. As rekor has captured in time, that a user in control of an OpenID account (lukehinds@gmail.com) was in possession of a certain key pair at a certain time. Another user can then get everything they need from rekor to verify the signature. 

The other advantage that comes with this, is that everyone can monitor the rekor for entries. So I may look out for people signing things as 'lukehinds@gmail.com" and if I see an instance, I know someone may have hacked my account.

So the whole big deal with keyless, is you don't have to look after the private key, or having someone hack into github and steal your key.

**Comment by lukehinds** [comment]
> > Eventually, we will be able to pass this certificate file to goreleaser artifacts, and expect 3 files after each release to use later using cosign to verify: checksum.txt, checksum.txt.sig, cosign.cert.

You should really use rekor for storing the artifacts or at least ensure the entries are in rekor.

**Comment by developer-guy** [comment]
Thank you so much, @lukehinds, for getting involved in this and sharing your thoughts with us; they are accommodating. However, IIAC, according to your sayings, we can't or shouldn't use these certificates for long-term verification, like storing it in the releases together with the binaries and sharing it with people to check our binary whether it is has been tampered with, right?

I know I asked it a couple of times because it might affect our development effort described [here]( https://github.com/goreleaser/goreleaser/issues/2659#issuecomment-965523990).

**Comment by shibumi** [comment]
> You should really use rekor for storing the artifacts or at least ensure the entries are in rekor.

IIRC, cosign uploads everything to rekor automatically. If not, it is possible to just add the `--rekor-url` parameter. I talked about this with @dlorenc and we came to the conclusion that having both (the transparency log in rekor AND a certificate on disk) might be useful.

> However, IIAC, according to your sayings, we can't or shouldn't use these certificates for long-term verification

As far as I know short-term certificates just mean that you validate if an artifact got signed in this specific timeframe. So yes. You totally can verify long-term artifacts. It's about the certificate, not about the artifact. With short-lived certificates you just ensure that the generated private key is short-lived... means: Nobody can retrieve a very old private key and sign artifacts with it and it's valid.. This is the problem that short-lived certificates are supposed to solve.

**Comment by shibumi** [comment]
@developer-guy

This is an example from my blog post from the 7th of november:

```
$ echo "hello world" > hello-world.txt
$ rekor-cli get --log-index 832324 --format json | jq -r '.Body.RekordObj.signature.content'  > hello-world.txt.sig
$  rekor-cli get --log-index 832324 --format json | jq -r '.Body.RekordObj.signature.publicKey.content' | base64 -d > pub.crt
$ COSIGN_EXPERIMENTAL=1 cosign verify-blob -cert pub.crt -signature hello-world.txt.sig hello-world.txt
No TUF root installed, using embedded CA certificate.
Certificate is trusted by Fulcio Root CA
Email: [chris@shibumi.dev]
Verified OK
tlog entry verified with uuid: "ee4b2e80cef72f8d4e3f00d695bd796ffd9e17b9cfd18ba96947493c4ee19f62" index: 832324
```

As you see.. the binary is still valid, although the certificate has already expired, because the signature got created in this short-lived time frame. :)

**Comment by developer-guy** [comment]
OMG! Everything is now clear 🙋🏻‍♂️ thank you so much @shibumi! That was super helpful information, so lucky to meet you 🤝

cc: @Dentrax @erkanzileli

Okay then, @caarlos, would you mind letting us to implement it to GoReleaser, or will you take care of it? We are so excited about the development, so, please feel free to assign it to us 🤝

**Comment by Dentrax** [comment]
Thanks for such a nice clarification Luke, and supporting the concept with a demonstration Christian! I clearly see now where we lost and thought in different way! 🧡

**Comment by caarlos0** [comment]
already working on it @developer-guy , will open the pr soon :D

**Comment by caarlos0** [comment]
check https://github.com/goreleaser/goreleaser/pull/2662

**Comment by lukehinds** [comment]
@developer-guy actually @shibumi blog describes it really well. Just an extra to what shibumi outlines, with github ODIC (which I believe you might be planning on using here), instead of the email, the CI run is captured 

![image](https://user-images.githubusercontent.com/7058938/141244521-ae310d4b-b491-4c42-bfd9-7c1367d5a2fe.png)

**Comment by shibumi** [comment]
@lukehinds this was one of my major questions about workload identities. Thank you! I always thought, that the OIDC ID from the project maintainer will transition to the workload identity. But letting the workload identity itself be the identity makes more sense.
:exploding_head: this makes sense now, too.

**Comment by lukehinds** [comment]
I have just been playing with `trimpath` and some other flags and we could have reproducible signed builds then (maybe goreleaser already does this)

This way you could compile a go binary (or wget it from somewhere), look up entries in rekor by the digest and see the same digest was signed by cosign in a github action.

**Comment by shibumi** [comment]
@lukehinds goreleaser is already reproducible (At least our Arch Linux release is reproducible. We are testing all packages for reproducibility):

![https://kurisu.shibumi.dev/share/img-2021-11-11-09-47-14.png](https://kurisu.shibumi.dev/share/img-2021-11-11-09-47-14.png)

**Comment by developer-guy** [comment]
> > Eventually, we will be able to pass this certificate file to goreleaser artifacts, and expect 3 files after each release to use later using cosign to verify: checksum.txt, checksum.txt.sig, cosign.cert.
> 
> I would suggest something like this:
> 
> ```yaml
> signs:
>   - cmd: cosign
>     signature: "${artifact}.sig"
>     certificate: "${artifact}.pem"
>     args: ["sign-blob", "-oidc-issuer=https://token.actions.githubusercontent.com", "-output=${signature}", "-output-pem": "${certificate}, "${artifact}"]
> ```

IIAC, @caarlos0, according to your PR#2662, this `certificate` field will be named as `pem`, right?

**Comment by caarlos0** [comment]
> IIAC, @caarlos0, according to your PR#2662, this `certificate` field will be named as `pem`, right? I have a small comment about it, [#2662 (comment)](https://github.com/goreleaser/goreleaser/pull/2662#discussion_r747363041)

bad prop rename, will be `certificate`

**Comment by developer-guy** [comment]
We are ready for the new releases 😋https://github.com/google/ko/pull/498/commits/06fb65dfd596444dd01c9b6f4750c76ee6981c16

**Comment by caarlos0** [comment]
noiceee

will likely release this later today, or in the worst case, over the weekend

**Comment by caarlos0** [comment]
> or in the worst case, over the weekend

more likely :|

**Comment by dlorenc** [comment]
Who releases the releaser!?!? :)

**Comment by shibumi** [comment]
I am sure we can release goreleaser with goreleaser + cosign in the future, too :D

**Comment by caarlos0** [comment]
It already signs with cosign, just not keyless - yet...

**Comment by shibumi** [comment]
@caarlos0 pretty sure @developer-guy  or me can donate a PR for this. The only question is if we should wait until the new production ready sigstore instances are ready. What do you think @dlorenc ? Does it make sense to sign keyless already with the "test" instances?

**Comment by dlorenc** [comment]
> @caarlos0 pretty sure @developer-guy or me can donate a PR for this. The only question is if we should wait until the new production ready sigstore instances are ready. What do you think @dlorenc ? Does it make sense to sign keyless already with the "test" instances?

I don't expect anything to change with the sigstore instances - they'll still be at the same hostnames and the data will all still be present. So feel free to use them now!

The messaging is more around cautioning people to not rely on them being 100% stable all the time until we finish up a bit more productionization work.

**Comment by developer-guy** [comment]
> @caarlos0 pretty sure @developer-guy or me can donate a PR for this. The only question is if we should wait until the new production ready sigstore instances are ready. What do you think @dlorenc ? Does it make sense to sign keyless already with the "test" instances?

there is an already PR for this @shibumi, and @caarlos0 already merged it to the main.
please see: https://github.com/goreleaser/goreleaser/pull/2662

**Comment by shibumi** [comment]
@developer-guy yep, but the github action workflow for keyless signatures for goreleaser is still missing. That's what I meant.

**Comment by developer-guy** [comment]
@caarlos0, does the v1.0.0 release of the GoReleaser include this issue? I couldn't find this issue in the release notes?

**Comment by developer-guy** [comment]
> @developer-guy yep, but the github action workflow for keyless signatures for goreleaser is still missing. That's what I meant.

@shibumi will you open another issue to track this?

**Comment by caarlos0** [comment]
I was planning on doing this once cosign have `--output-certificate` or similar, so we can still upload the pub key to the release and users can still validate the signatures on air-gapped environments...
